File: 1418-fair-distribution-of-cookies/fair-distribution-of-cookies.java
class Solution {

    int minUnfair = Integer.MAX_VALUE;
    void result(int[] cookies , int[] child , int idx , int k) {

        // base case
        if(idx==cookies.length) {

            int max = 0;
            for (int i = 0; i < child.length; i++) {
                max = Math.max(max, child[i]);
            }

            minUnfair = Math.min(minUnfair , max);        
            return;    

        }
        

        for(int i = 0 ; i<k ; i++) {
            child[i] = child[i] + cookies[idx];

            // explore
            result(cookies , child , idx+1 , k);

            //backtrack
            child[i] = child[i] - cookies[idx];

            if(child[i]==0) {
                break;
            }
        }
    }

    public int distributeCookies(int[] cookies, int k) {
        int n = cookies.length;
        int[] child = new int[k];

        result(cookies , child , 0 , k);
        return minUnfair;
    }
}

File: 1418-fair-distribution-of-cookies/fair-distribution-of-cookies.cpp
class Solution {
    int minUnfair = INT_MAX;

    void result(vector<int>& cookies, vector<int>& child, int idx, int k) {
        // base case
        if (idx == cookies.size()) {
            int maxVal = 0;
            for (int i = 0; i < k; ++i) {
                maxVal = max(maxVal, child[i]);
            }
            minUnfair = min(minUnfair, maxVal);
            return;
        }

        for (int i = 0; i < k; ++i) {
            child[i] += cookies[idx];

            // explore
            result(cookies, child, idx + 1, k);

            // backtrack
            child[i] -= cookies[idx];

            if (child[i] == 0) {
                break; // pruning to avoid symmetric states
            }
        }
    }

public:
    int distributeCookies(vector<int>& cookies, int k) {
        vector<int> child(k, 0);
        result(cookies, child, 0, k);
        return minUnfair;
    }
};
